### DISCUSSION TITLE
Guidance for bugfix in swc's `chain!()` of `ImportAnalyzer` and `CommonJs` transformers

### DISCUSSION THREAD

**Discussion by cspotcode** [swc-project/swc#1826] [General]
I'm attempting a bugfix for #1786.  The README suggests that I should ask here for guidance understanding swc's codebase.

I see that `ImportAnalyzer` and `CommonJs` transformers both use `scope.import_types` to understand if a `require()` needs to be wrapped in `interopRequireDefault` or `interopRequireWildcard`.  However, the transformers do not seem to share a `Scope` instance.

This causes problems when `ImportAnalyzer` marks an import as wildcard, but `CommonJs` does not know this and treats it as a default import.

Should transformers share a single `scope` instance, or should `import_types` be copied from one transformer to the next?

**Comment by kdy1** [comment]
Sharing scope is better, but it may require some strange tricks or `Mutex`  (Possibly `RwLock`).

**Reply by cspotcode** [reply]
I was trying to understand that.  If the transformers run sequentially, then two are never running at the same time, right?

If so, can we use `RefCell` to give a mutable borrow of the `Scope` to each transformer while it is running?

I think this will require augmenting the `chain!()` so that it owns the `Scope`.  The transformers can be wrapped in some sort of factory which creates them when they execute and drops immediately after, instead of creating them all ahead-of-time.

Is this a smart idea?

**Reply by kdy1** [reply]
You can do it with `Rc<RefCell>>` and pass down the rc to both pass.

> I think this will require augmenting the chain!() so that it owns the Scope. The transformers can be wrapped in some sort of factory which creates them when they execute and drops immediately after, instead of creating them all ahead-of-time.

It's possible, but it should allow injecting passes between analysis and import replacer.
